src/Columns/IColumn.cpp

```diff
@@ -499,12 +500,43 @@ static void fillColumnFromRowRefs(ColumnType * col, const DataTypePtr & type, co
             if constexpr (row_refs_are_ranges)
             {
                 row_ref_list->assertIsRange();
-                col->insertRangeFrom(*(*row_ref_list->columns)[source_column_index_in_block], row_ref_list->row_num, row_ref_list->rows);
+                const auto & source_column = (*row_ref_list->columns)[source_column_index_in_block];
+                if (source_columns_might_be_replicated && source_column->isReplicated())
+                {
+                    const auto & source_replicated = assert_cast<const ColumnReplicated &>(*source_column);
+                    const auto & source_nested_column = source_replicated.getNestedColumn();
+                    const auto & source_indexes = source_replicated.getIndexes();
+                    for (size_t i = row_ref_list->row_num; i != row_ref_list->row_num + row_ref_list->rows; ++i)
+                        col->insertFrom(*source_nested_column, source_indexes.getIndexAt(i));
```

Comment:
But should not internal implementation of COlumnReplicated::insertRangeFrom be the same way efficient?

Reply:
Column we insert to is not ColumnReplicated. It can happen when column from right table in JOIN is ColumnReplicated. It would be great to make column we insert to also `ColumnReplicated`. We can do it if we can check if right table column is replicated during result column creation, but I couldn't find the place where we cn do it (please help, code of JOINs is quite complex)